File: app/src/main/java/com/cs407/virtualwiscard/MyHostApduService.java
package com.cs407.virtualwiscard;

import android.content.Context;
import android.content.SharedPreferences;
import android.nfc.cardemulation.HostApduService;
import com.google.firebase.firestore.FirebaseFirestore;
import android.os.Bundle;
import android.util.Log;

public class MyHostApduService extends HostApduService {

    private static final String SELECT_APDU_HEADER = "00A40400"; // APDU command for SELECT
    private static final String RESPONSE_ERROR = "6F00"; // Error response

    @Override
    public byte[] processCommandApdu(byte[] apdu, Bundle extras) {
        Log.d("HCE", "Received APDU: " + bytesToHex(apdu));
        String apduHex = bytesToHex(apdu);

        /*
        if (apduHex.startsWith(SELECT_APDU_HEADER)) {
            // Handle the SELECT command
            SharedPreferences sharedPreferences = getSharedPreferences("appPrefs", Context.MODE_PRIVATE);
            String wiscardNumber = sharedPreferences.getString("wiscardNumberForNFC", "Unknown Wiscard Number");
            byte[] responseBytes = wiscardNumber.getBytes();
            Log.d("HCE", "APDU response: " + bytesToHex(responseBytes));
            return responseBytes;
        }
        */

        if (apduHex.startsWith(SELECT_APDU_HEADER)) {
            Log.d("HCE", "AID selected");

            // Check user permissions from cache
            boolean userHasAccess = checkCachedUserAccess();

            String response = "";
            if (userHasAccess) {
                Log.d("HCE", "Permission granted: PASS");
                response = "PASS"; // User has permission
            } else {
                Log.d("HCE", "Permission denied: FAIL");
                response = "FAIL"; // User does not have permission
            }

            byte[] responseBytes = response.getBytes();
            Log.d("HCE", "APDU response: " + bytesToHex(responseBytes));
            return responseBytes;
        }

        // Return error if command is not recognized
        return hexStringToByteArray(RESPONSE_ERROR);
    }

    @Override
    public void onDeactivated(int reason) {
        Log.d("HCE", "onDeactivated called with reason: " + reason);
        switch (reason) {
            case DEACTIVATION_LINK_LOSS:
                Log.d("HCE", "Deactivation due to link loss.");
                break;
            case DEACTIVATION_DESELECTED:
                Log.d("HCE", "Deactivation due to deselection.");
                break;
            default:
                Log.d("HCE", "Deactivation for unknown reason.");
        }
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder result = new StringBuilder();
        for (byte b : bytes) {
            result.append(String.format("%02X", b));
        }
        return result.toString();
    }

    private byte[] hexStringToByteArray(String s) {
        int len = s.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                    + Character.digit(s.charAt(i + 1), 16));
        }
        return data;
    }

    // Add Kotlin equivalent function in Java
    public interface Callback {
        void onComplete(boolean success);
    }

    public static void updateUserAccessCache(Context context, String username, Callback callback) {
        FirebaseFirestore firestore = FirebaseFirestore.getInstance();
        firestore.collection("users").document(username)
                .get()
                .addOnSuccessListener(document -> {
                    boolean hasAccess = document != null && document.getBoolean("hasAccess") != null && document.getBoolean("hasAccess");
                    Log.d("MyHostApduService", "Access fetched for " + username + ": " + hasAccess);
                    SharedPreferences sharedPreferences = context.getSharedPreferences("user_prefs", Context.MODE_PRIVATE);
                    sharedPreferences.edit().putBoolean("has_access", hasAccess).apply();
                    callback.onComplete(true); // Notify success
                })
                .addOnFailureListener(e -> {
                    Log.e("MyHostApduService", "Failed to fetch access for " + username, e);
                    callback.onComplete(false); // Notify failure
                });
    }

    public boolean checkCachedUserAccess() {
        SharedPreferences sharedPreferences = getSharedPreferences("user_prefs", Context.MODE_PRIVATE);
        boolean hasAccess = sharedPreferences.getBoolean("has_access", false);
        Log.d("MyHostApduService", "Cached user access: " + hasAccess);
        return hasAccess;
    }
}

File: app/src/main/java/com/cs407/virtualwiscard/MyHostApduService.kt
package com.example.capp

import android.nfc.cardemulation.HostApduService
import android.os.Bundle
import android.util.Log

class MyHostApduService : HostApduService() {
    override fun processCommandApdu(apdu: ByteArray, extras: Bundle): ByteArray {
        Log.d("HCE", "Received APDU: " + bytesToHex(apdu))
        val apduHex = bytesToHex(apdu)

        if (apduHex.startsWith(SELECT_APDU_HEADER)) {
            // Handle the SELECT command
            val message = "Hello NFC!"
            val responseBytes = message.toByteArray()
            Log.d("HCE", "APDU response: " + bytesToHex(responseBytes))
            return buildResponse(message.toByteArray())
        }

        // Return error if command is not recognized
        return hexStringToByteArray(RESPONSE_ERROR)
    }

    override fun onDeactivated(reason: Int) {
        Log.d("HCE", "onDeactivated called with reason: $reason")
        when (reason) {
            DEACTIVATION_LINK_LOSS -> Log.d("HCE", "Deactivation due to link loss.")
            DEACTIVATION_DESELECTED -> Log.d("HCE", "Deactivation due to deselection.")
            else -> Log.d("HCE", "Deactivation for unknown reason.")
        }
    }

    private fun buildResponse(data: ByteArray): ByteArray {
        return concatenateArrays(data, hexStringToByteArray(RESPONSE_OK))
    }

    private fun concatenateArrays(array1: ByteArray, array2: ByteArray): ByteArray {
        val result = ByteArray(array1.size + array2.size)
        System.arraycopy(array1, 0, result, 0, array1.size)
        System.arraycopy(array2, 0, result, array1.size, array2.size)
        return result
    }

    private fun bytesToHex(bytes: ByteArray): String {
        val result = StringBuilder()
        for (b in bytes) {
            result.append(String.format("%02X", b))
        }
        return result.toString()
    }

    private fun hexStringToByteArray(s: String): ByteArray {
        val len = s.length
        val data = ByteArray(len / 2)
        var i = 0
        while (i < len) {
            data[i / 2] = ((s[i].digitToIntOrNull(16) ?: -1 shl 4)
            + s[i + 1].digitToIntOrNull(16)!! ?: -1).toByte()
            i += 2
        }
        return data
    }

    companion object {
        private const val SELECT_APDU_HEADER = "00A40400" // APDU command for SELECT
        private const val RESPONSE_OK = "9000" // Success response
        private const val RESPONSE_ERROR = "6F00" // Error response

        private const val KEEP_ALIVE_APDU = "00" // Empty APDU for keeping connection alive
    }
}